lib/libc/sys/kenv.2
.\"
.\" Copyright (C) 2002 Chad David <davidc@FreeBSD.org>. All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice(s), this list of conditions and the following disclaimer as
.\"    the first lines of this file unmodified other than the possible
.\"    addition of one or more copyright notices.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice(s), this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDER(S) ``AS IS'' AND ANY
.\" EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
.\" WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
.\" DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT HOLDER(S) BE LIABLE FOR ANY
.\" DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
.\" (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
.\" SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
.\" CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
.\" DAMAGE.
.\"
.\" $FreeBSD: head/lib/libc/sys/kenv.2 267774 2014-06-23 08:25:03Z bapt $
.\"
.Dd March 19, 2016
.Dt KENV 2
.Os
.Sh NAME
.Nm kenv
.Nd kernel environment
.Sh LIBRARY
.Lb libc
.Sh SYNOPSIS
.In kenv.h
.Ft int
.Fn kenv "int action" "const char *name" "char *value" "int len"
.Sh DESCRIPTION
The
.Fn kenv
system call manipulates kernel environment variables.
It supports the well known userland actions of getting, setting and unsetting
environment variables, as well as the ability to dump all of the entries in
the kernel environment.
.Pp
The
.Fa action
argument can be one of the following:
.Bl -tag -width ".Dv KENV_UNSET"
.It Dv KENV_GET
Returns the value associated with the named kernel environment variable.
If the variable is not found, \-1 is returned and
the global variable
.Va errno
is set to
.Er ENOENT .
Only the number of bytes available in
.Fa value
are copied out.
.It Dv KENV_SET
Sets or adds a new kernel environment variable.
This option is only available to the superuser.
.It Dv KENV_UNSET
Unsets the kernel environment variable
.Fa name .
If the variable does not exist, \-1 is returned and
the global variable
.Va errno
is set to
.Er EINVAL .
This option is only available to the superuser.
.It Dv KENV_DUMP
Dumps as much of the kernel environment as will fit in
.Fa value .
If
.Fa value
is
.Dv NULL ,
.Fn kenv
will return the number of bytes required to copy out the entire environment.
.El
.Pp
The
.Fa name
argument is the name of the environment variable to be affected.
In the case of
.Dv KENV_DUMP
it is ignored.
.Pp
The
.Fa value
argument contains either the value to set the environment variable
.Fa name
to in the case of
.Dv KENV_SET ,
or it points to the location where
.Fn kenv
should copy return data to in the case of
.Dv KENV_DUMP
and
.Dv KENV_GET .
If
.Fa value
is
.Dv NULL
in the case of
.Dv KENV_DUMP ,
.Fn kenv
will return the number of bytes required to copy out the entire environment.
.Pp
The
.Fa len
argument indicates how many bytes of storage
.Fa value
points to.
.Sh RETURN VALUES
The
.Fn kenv
system call returns 0 if successful in the case of
.Dv KENV_SET
and
.Dv KENV_UNSET ,
and the number of bytes copied into
.Fa value
in the case of
.Dv KENV_DUMP
and
.Dv KENV_GET .
If an error occurs, a value of \-1 is returned and
the global variable
.Va errno
is set to indicate the error.
.Sh ERRORS
The
.Fn kenv
system call
will fail if:
.Bl -tag -width Er
.It Bq Er EINVAL
The
.Fa action
argument
is not a valid option, or the length of the
.Fa value
is less than 1 for a
.Dv KENV_SET .
.It Bq Er ENOENT
no value could be found for
.Fa name
for a
.Dv KENV_SET
or
.Dv KENV_UNSET .
.It Bq Er EPERM
a user other than the superuser attempted to set or unset a kernel
environment variable.
.It Bq Er EFAULT
bad address was encountered while attempting to copy in user arguments,
or copy out value(s).
.It Bq Er ENAMETOOLONG
the name of a variable supplied by the user is longer than
.Dv KENV_MNAMELEN
or the value of a variable is longer than
.Dv KENV_MVALLEN .
.El
.Sh SEE ALSO
.Xr kenv 1
.Sh AUTHORS
.An -nosplit
This manual page was written by
.An Chad David Aq Mt davidc@FreeBSD.org .
.Pp
The
.Fn kenv
system call was written by
.An Maxime Henrion Aq Mt mux@FreeBSD.org .


usr.bin/kenv/kenv.c
/*-
 * Copyright (c) 2000  Peter Wemm <peter@freebsd.org>
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHORS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHORS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 * $FreeBSD: head/bin/kenv/kenv.c 250666 2013-05-15 18:38:28Z delphij $
 */

#include <sys/types.h>
#include <sys/sysctl.h>
#include <err.h>
#include <kenv.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

static void	usage(void);
static int	kdumpenv(void);
static int	kgetenv(const char *);
static int	ksetenv(const char *, char *);
static int	kunsetenv(const char *);

static int hflag = 0;
static int Nflag = 0;
static int qflag = 0;
static int uflag = 0;
static int vflag = 0;

static void
usage(void)
{
	(void)fprintf(stderr, "%s\n%s\n%s\n",
	    "usage: kenv [-hNq]",
	    "       kenv [-qv] variable[=value]",
	    "       kenv [-q] -u variable");
	exit(1);
}

int
main(int argc, char **argv)
{
	char *env, *eq, *val;
	int ch, error;

	error = 0;
	val = NULL;
	env = NULL;
	while ((ch = getopt(argc, argv, "hNquv")) != -1) {
		switch (ch) {
		case 'h':
			hflag++;
			break;
		case 'N':
			Nflag++;
			break;
		case 'q':
			qflag++;
			break;
		case 'u':
			uflag++;
			break;
		case 'v':
			vflag++;
			break;
		default:
			usage();
		}
	}
	argc -= optind;
	argv += optind;
	if (argc > 0) {
		env = argv[0];
		eq = strchr(env, '=');
		if (eq != NULL) {
			*eq++ = '\0';
			val = eq;
		}
		argv++;
		argc--;
	}
	if ((hflag || Nflag) && env != NULL)
		usage();
	if (argc > 0 || ((uflag || vflag) && env == NULL))
		usage();
	if (env == NULL) {
		error = kdumpenv();
		if (error && !qflag)
			warn("kdumpenv");
	} else if (val == NULL) {
		if (uflag) {
			error = kunsetenv(env);
			if (error && !qflag)
				warnx("unable to unset %s", env);
		} else {
			error = kgetenv(env);
			if (error && !qflag)
				warnx("unable to get %s", env);
		}
	} else {
		error = ksetenv(env, val);
		if (error && !qflag)
			warnx("unable to set %s to %s", env, val);
	}
	return (error);
}

static int
kdumpenv(void)
{
	char *buf, *cp;
	int buflen, envlen;

	envlen = kenv(KENV_DUMP, NULL, NULL, 0);
	if (envlen < 0)
		return (-1);
	for (;;) {
		buflen = envlen * 120 / 100;
		buf = malloc(buflen + 1);
		if (buf == NULL)
			return (-1);
		memset(buf, 0, buflen + 1);	/* Be defensive */
		envlen = kenv(KENV_DUMP, NULL, buf, buflen);
		if (envlen < 0) {
			free(buf);
			return (-1);
		}
		if (envlen > buflen)
			free(buf);
		else
			break;
	}

	for (; *buf != '\0'; buf += strlen(buf) + 1) {
		if (hflag) {
			if (strncmp(buf, "hint.", 5) != 0)
				continue;
		}
		cp = strchr(buf, '=');
		if (cp == NULL)
			continue;
		*cp++ = '\0';
		if (Nflag)
			printf("%s\n", buf);
		else
			printf("%s=\"%s\"\n", buf, cp);
		buf = cp;
	}
	return (0);
}

static int
kgetenv(const char *env)
{
	char buf[1024];
	int ret;

	ret = kenv(KENV_GET, env, buf, sizeof(buf));
	if (ret == -1)
		return (ret);
	if (vflag)
		printf("%s=\"%s\"\n", env, buf);
	else
		printf("%s\n", buf);
	return (0);
}

static int
ksetenv(const char *env, char *val)
{
	int ret;

	ret = kenv(KENV_SET, env, val, strlen(val)+1);
	if (ret == 0)
		printf("%s=\"%s\"\n", env, val);
	return (ret);
}

static int
kunsetenv(const char *env)
{
	int ret;

	ret = kenv(KENV_UNSET, env, NULL, 0);
	return (ret);
}
